File: android/DDSChat/src/main/java/org/opensplice/mobile/demo/ddschat/MainActivity.java
package org.opensplice.mobile.ddschat;

import java.util.concurrent.TimeoutException;

import android.os.Bundle;
import android.os.Handler;
import android.app.Activity;
import android.util.Log;
import android.view.Menu;
import android.view.View;
import android.widget.*;

import org.omg.dds.core.event.DataAvailableEvent;
import org.omg.dds.sub.DataReader;
import org.omg.dds.sub.Sample;
import org.omg.dds.sub.Sample.Iterator;
import org.omg.dds.sub.SampleState;
import org.opensplice.mobile.demo.idl.ChatMessage;

public class MainActivity extends Activity {
	
	private static final String user = Config.defaultUserName;
	private final Handler handler = new Handler();
	private ArrayAdapter<String> chatMessages;

	public class ChatMessageListener extends ChatMessageDataListener {

		private DataReader<ChatMessage> dr;

		public ChatMessageListener() {
			ChatApplication app = (ChatApplication) getApplication();
			dr = app.reader();
		}

		@Override
		public void onDataAvailable(DataAvailableEvent<ChatMessage> dae) {
			final Iterator<ChatMessage> i = dr.read();

			Log.i(Config.MA_TAG, ">>> DataReaderListener.onDataAvailable");
			if (i.hasNext()) {
				Runnable dispatcher = new Runnable() {
					public void run() {
						while (i.hasNext()) {
							Sample<ChatMessage> s = i.next();

							if (s.getSampleState() == SampleState.NOT_READ) {
								ChatMessage cm = s.getData();
								chatMessages.add(cm.user + " >  " + cm.msg);
							}
						}						
					}
				};
				handler.post(dispatcher);
			}
		}
	
	}

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_main);
		chatMessages = new ArrayAdapter<String>(this, R.layout.messages);
		chatMessages.add("Welcome to the DDS Chat Room");
		ListView mview = (ListView) findViewById(R.id.messageList);
		mview.setAdapter(chatMessages);
		ChatApplication app = (ChatApplication) getApplication();

		app.reader().setListener(new ChatMessageListener());

	}

	@Override
	public boolean onCreateOptionsMenu(Menu menu) {
		// Inflate the menu; this adds items to the action bar if it is present.
		getMenuInflater().inflate(R.menu.main, menu);
		return true;
	}

	public void sendChatMessage(View view) {
		EditText editText = (EditText) findViewById(R.id.message);
		String msg = editText.getText().toString();
		editText.setText("");
		ChatApplication app = (ChatApplication) getApplication();
		try {
			Log.i(Config.MA_TAG, ">>> Sending data " + msg);
			app.writer().write(new ChatMessage(user, msg));
		} catch (TimeoutException te) {
		}
	}

}

File: android/Oximeter/src/main/scala/org/opensplice/mobile/demo/MainActivity.scala
package org.opensplice.mobile.demo

import android.app.Activity
import android.os.{ Bundle, Message, Handler }
import android.view.Menu
import android.widget.TextView
import android.util.Log
import android.graphics._

import scala.language.implicitConversions
import org.opensplice.mobile.demo.prelude._
import scala.collection.JavaConversions._
import org.omg.dds.sub.Sample

import com.androidplot.Plot

import com.androidplot.xy._
import com.androidplot.series._
import com.androidplot.xy._

import java.text._
import java.util.Arrays
import java.util.Date

class MainActivity extends Activity {

  var did: TextView = null
  var spO2: TextView = null
  var ppm: TextView = null
  var bpm: TextView = null
  var pleth: TextView = null

  var dynamicPlot: XYPlot = null
  var plethSeries: OximeterTimeSeries = null
  val handler = new Handler()

  val activityUpdater = new Runnable {
    def run() {

      plethSeries.refresh()
      val app = getApplication().asInstanceOf[OximeterApp]
      val odr = app.endpoints.oximetryDR
      val oximetry = odr.read()
      oximetry.find(_ => true) map { s =>
        val d = s.getData()
        did.setText(d.deviceId)
        spO2.setText(d.spO2.toString)
        bpm.setText(d.bpm.toString)
        pleth.setText(d.pleth.toString)
        ppm.setText(d.rr.toString)

      }
    }
  }

  val scheduler = new Thread(() => {
    var interrupted = false
    while (!interrupted) {
      interrupted =
        try {
          handler.post(activityUpdater)
          Thread.sleep(AppConfig.refreshPeriod)
          false
        } catch {
          case ie: InterruptedException => {
            true
          }

        }
    }
  })

  import scala.collection.JavaConversions._
  import scala.collection.JavaConverters._

  override def onCreate(savedInstanceState: Bundle) {

    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_main)
    dynamicPlot = get[XYPlot](R.id.XYPlot)
    dynamicPlot.setTitle("Real-Time Pleth")
    // only display whole numbers in domain labels
    dynamicPlot.getGraphWidget().setDomainValueFormat(new DecimalFormat("0"));

    val app = getApplication().asInstanceOf[OximeterApp]
    val pdr = app.endpoints.plethDR
    plethSeries = new OximeterTimeSeries(AppConfig.timeSeriesTitle, pdr, dynamicPlot)

    dynamicPlot.addSeries(plethSeries,
      new LineAndPointFormatter(Color.rgb(0x33, 0x66, 0x99),
        null, //Color.rgb(0x33, 0x66, 0x99), 
        Color.rgb(0x66, 0x99, 0xff)))

    dynamicPlot.setDomainStepMode(XYStepMode.SUBDIVIDE);

    did = get[TextView](R.id.deviceId)
    spO2 = get[TextView](R.id.spO2)
    ppm = get[TextView](R.id.bpm)
    bpm = get[TextView](R.id.bpm)
    pleth = get[TextView](R.id.pleth)

    scheduler.start()
  }

  override def onCreateOptionsMenu(menu: Menu) = {
    super.onCreateOptionsMenu(menu)
    getMenuInflater().inflate(R.menu.main, menu)
    true
  }

  override def onDestroy() {    
    super.onDestroy()
    scheduler.interrupt()   
  }

  def get[T](id: Int) = findViewById(id).asInstanceOf[T]

}